substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -122,66 +299,137 @@ where
 	}
 }
 
-impl<Address, Call, Signature, Extra: SignedExtension>
-	UncheckedExtrinsic<Address, Call, Signature, Extra>
-{
-	/// New instance of a signed extrinsic aka "transaction".
-	pub fn new_signed(function: Call, signed: Address, signature: Signature, extra: Extra) -> Self {
-		Self { signature: Some((signed, signature, extra)), function }
+impl<Address, Call, Signature, Extension> UncheckedExtrinsic<Address, Call, Signature, Extension> {
+	/// New instance of a bare (ne unsigned) extrinsic. This could be used for an inherent or an
+	/// old-school "unsigned transaction" (which are new being deprecated in favour of general
+	/// transactions).
+	#[deprecated = "Use new_bare instead"]
+	pub fn new_unsigned(function: Call) -> Self {
+		Self::new_bare(function)
 	}
 
-	/// New instance of an unsigned extrinsic aka "inherent".
-	pub fn new_unsigned(function: Call) -> Self {
-		Self { signature: None, function }
+	/// Returns `true` if this extrinsic instance is an inherent, `false`` otherwise.
+	pub fn is_inherent(&self) -> bool {
+		matches!(self.preamble, Preamble::Bare(_))
 	}
-}
 
-impl<Address: TypeInfo, Call: TypeInfo, Signature: TypeInfo, Extra: SignedExtension + TypeInfo>
-	Extrinsic for UncheckedExtrinsic<Address, Call, Signature, Extra>
-{
-	type Call = Call;
+	/// Returns `true` if this extrinsic instance is an old-school signed transaction, `false`
+	/// otherwise.
+	pub fn is_signed(&self) -> bool {
+		matches!(self.preamble, Preamble::Signed(..))
+	}
 
-	type SignaturePayload = UncheckedSignaturePayload<Address, Signature, Extra>;
+	/// Create an `UncheckedExtrinsic` from a `Preamble` and the actual `Call`.
+	pub fn from_parts(function: Call, preamble: Preamble<Address, Signature, Extension>) -> Self {
+		Self { preamble, function }
+	}
 
-	fn is_signed(&self) -> Option<bool> {
-		Some(self.signature.is_some())
+	/// New instance of a bare (ne unsigned) extrinsic.
+	pub fn new_bare(function: Call) -> Self {
+		Self { preamble: Preamble::Bare(EXTRINSIC_FORMAT_VERSION), function }
 	}
 
-	fn new(function: Call, signed_data: Option<Self::SignaturePayload>) -> Option<Self> {
-		Some(if let Some((address, signature, extra)) = signed_data {
-			Self::new_signed(function, address, signature, extra)
-		} else {
-			Self::new_unsigned(function)
-		})
+	/// New instance of a bare (ne unsigned) extrinsic on extrinsic format version 4.
+	pub fn new_bare_legacy(function: Call) -> Self {
+		Self { preamble: Preamble::Bare(LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION), function }
+	}
+
+	/// New instance of an old-school signed transaction.
+	pub fn new_signed(
+		function: Call,
+		signed: Address,
+		signature: Signature,
+		tx_ext: Extension,
+	) -> Self {
+		Self {
+			preamble: Preamble::Signed(
+				signed,
+				signature,
+				EXTENSION_VERSION,
+				tx_ext,
+				EXTRINSIC_FORMAT_VERSION,
+			),
+			function,
+		}
+	}
+
+	/// New instance of an old-school signed transaction on extrinsic format version 4.
+	pub fn new_signed_legacy(
```

Comment:
Do we need this? Probably for tests? If yes, can we put it behind `cfg(test)`?

Reply:
> Do we need this? Probably for tests? If yes, can we put it behind `cfg(test)`?

I think we will need it for the P/K bridge relayer, because Polkadot and Kusama runtime upgrades are asynchronous. At some point, KusamaBridgeHub will be upgraded to the new format (5), while PolkadotBridgeHub will still use the older format (4). Our relayer will need to support both `new_signed` and `new_signed_legacy`, for example [here](https://github.com/paritytech/parity-bridges-common/blob/master/relay-clients/client-bridge-hub-polkadot/src/lib.rs#L125-L130).

Can we deprecate it and remove it by Q1-Q2/25?

Reply:
Yes, we should probably just mark it as deprecated until all of the tooling migrates to v5.